### ISSUE TITLE
Parsing string interpolation

### ISSUE THREAD

**Issue by SupraSummus** [dhall-lang/dhall-lang#280]
I found a problem with dhall-haskell implementation and decided to open an issue here, because discussion comes to dhall standard usually (anticipating from previous cases).

Expression is `"${ not_really_an_expression ;-) }"`
 * dhall standard allows it
 * dhall-haskell rejects it on parsing stage

I think we should forbid using bare `$` in quoted strings. Only escaped version (`\$`) should be allowed.

**Comment by SupraSummus** [comment]
Similar problem occurs for single quoted strings and for block comments. Current behaviour described by standard is:

`''${ not_an_expression ;-) }''` is just a string without interpolation

`{- {- -} 1` evaluates to `1`, because the block comment encloses ` {- `

In both cases dhall-haskell returns parse error.

**Comment by Gabriel439** [comment]
Just to make sure I understand: the issue is that the standard interprets `"…${x}…"` differently depending on whether or not `x` is a syntactically valid Dhall expression.  If `x` is syntactically valid then it is parsed as string interpolation and if `x` is not syntactically valid then it is parsed as the actual contents of the string and not as an interpolation.

I would hesitate to make `$` invalid within a string literal since that might be a bit jarring for users, but what about making `${` invalid within a double-quoted string literal?  In other words, instead of requiring users to escape all `$` signs we can instead only require them to escape the two-character sequence `${`.  I haven't yet checked how easy that would be to standardize, though.

**Comment by SupraSummus** [comment]
Yes, your description of the issue matches with what I had in mind.

Making unescaped `${` invalid seems like a good solution. The escape sequence would be `\${`?

**Comment by Gabriel439** [comment]
@SupraSummus: To be precise, there would not be a new `\${` escape sequence for `${`.  There would still just be a `\$` escape sequence for `$`, which just happens to mean that `\${` is still interpreted as `${`.

However, it all depends on how difficult it is to specify a grammar for double-quoted string literals that disallows an unescaped `${`.  If it's too tricky then we could fall back to disallowing an unescaped `$`.
